src/device/pci/rings.rs

```diff
@@ -652,6 +704,90 @@ mod tests {
         }
     }
 
+    fn dummy_trb() -> EventTrb {
+        EventTrb::new_transfer_event_trb(
+            0,                       // trb_pointer
+            0,                       // trb_transfer_length
+            CompletionCode::Success, // completion_code
+            false,                   // event_data
+            1,                       // endpoint_id
+            1,                       // slot_id
+        )
+    }
+
+    fn assert_trb_written(ram: &TestingRamDevice, addr: u64) {
+        let mut buf = [0u8; 16];
+        ram.read_bulk(addr, &mut buf);
+        assert_ne!(buf, [0u8; 16], "TRB not written at address {:#x}", addr);
+    }
```

Comment:
You should check the cycle bit instead of the non-zero check. Your approach works fine with zero-initialized memory and no wrap-around. After the wrap-around, you enqueue the dummy TRB, which overwrites the old dummy TRB: The only effect is that the cycle bit changes from 1 to 0. Your assertion succeeds because you see the completion code and ids, but it would also succeed without the wrap-around write. The only way to check is to confirm that after the wrap-around write the cycle bit was set to 0. 
```suggestion
    fn assert_trb_written(ram: &TestingRamDevice, addr: u64, cycle_state: bool) {
        let mut buf = [0u8; 16];
        ram.read_bulk(addr, &mut buf);
        let cycle_bit = buf[12] & 0x1 != 0;
        assert_eq(cycle_bit, cycle_state, "TRB not written at address {:#x}", addr);
    }
```

Reply:
Just saw below that you don't actually check if an TRB was written after wrap-around. I think you should, so this change is still relevant.